webpack/assets/javascripts/react_app/components/ColumnSelector/ColumnSelector.js

```diff
@@ -0,0 +1,200 @@
+      }
+    });
+    return result;
+  };
+
+  const toggleModal = () => {
+    setSelectedColumns(initialColumns);
+    setModalOpen(!isModalOpen);
+  };
+
+  const isDisabled = () => {
+    const hasPartialCheck = selectedColumns.map(
+      item => item.checkProps.checked
+    );
+    return hasPartialCheck.every(el => el === false);
+  };
+
+  const updateCheckBox = (treeViewItem, checked = true) => {
+    treeViewItem.checkProps.checked = checked;
+    if (treeViewItem.children) {
+      treeViewItem.children.forEach(item => {
+        item.checkProps.checked = checked;
+      });
+    }
+  };
+
+  const onCheck = (evt, treeViewItem) => {
+    const { checked } = evt.target;
+    const checkedItemTree = selectedColumns
+      .map(column => Object.assign({}, column))
+      .filter(item => filterItems(item, treeViewItem));
+    const flatCheckedItems = flattenTree(checkedItemTree);
+
+    if (checked) {
+      updateCheckBox(treeViewItem);
+      setSelectedColumns(
+        selectedColumns
+          .concat(
+            flatCheckedItems.filter(
+              item => !selectedColumns.some(i => i.key === item.key)
+            )
+          )
+          .filter(item => item.children)
+      );
+    } else {
+      updateCheckBox(treeViewItem, false);
+      setSelectedColumns(
+        selectedColumns.filter(item =>
+          flatCheckedItems.some(i => i.key === item.key)
+        )
+      );
+    }
+    selectedColumns.map(category => areDescendantsChecked(category));
+  };
+
+  const isChecked = dataItem => dataItem.checkProps.checked;
+  const areDescendantsChecked = dataItem => {
+    if (dataItem.children) {
+      if (dataItem.children.every(child => isChecked(child))) {
+        dataItem.checkProps.checked = true;
+      } else if (dataItem.children.some(child => isChecked(child))) {
+        dataItem.checkProps.checked = null;
+      } else {
+        dataItem.checkProps.checked = false;
+      }
+    }
+  };
+
+  if (!tablePreference) {
+    updateTablePreference();
+  }
+
+  return (
+    <div className="pf-c-select-input">
+      <div className="pf-c-input-group" id="column-selector">
+        <Button
+          id="btn-select-columns"
+          variant="link"
+          icon={<ColumnsIcon />}
+          iconPosition="left"
+          className="columns-selector"
+          onClick={() => toggleModal()}
+        >
+          {__('Manage columns')}
+        </Button>
+        <Modal
+          variant={ModalVariant.small}
+          title={__('Manage columns')}
+          isOpen={isModalOpen}
+          onClose={toggleModal}
+          tabIndex={0}
+          description={__('Select columns to display in the table.')}
+          position="top"
+          actions={[
+            <Button
+              key="save"
+              variant="primary"
+              isDisabled={isDisabled()}
+              onClick={() => {
+                updateTablePreference();
+                changeQuery({});
```

Comment:
This should be part of `updateTablePreference` function, otherwise it will behave asynchronously.
```suggestion
```